feature-governance-impl/src/main/java/io/novafoundation/nova/feature_governance_impl/domain/referendum/vote/RealVoteReferendumInteractor.kt

```diff
@@ -149,130 +152,19 @@ class RealVoteReferendumInteractor(
-            ),
-            transferableChange = Change(
-                previousValue = currentTransferablePlanks,
-                newValue = newTransferablePlanks,
-            )
-        )
-    }
-
-    override suspend fun reusableLocks(): List<ReusableLock> {
-        return buildList {
-            addIfPositive(ReusableLock.Type.GOVERNANCE, currentMaxGovernanceLocked)
-            addIfPositive(ReusableLock.Type.ALL, allMaxLocked)
-        }
-    }
 
-    private fun estimateUnlocksAt(changedVote: AccountVote.Standard?): BlockNumber {
-        val priorUnlocksAt = priorUnlocksAt()
-        val votesEstimatedUnlocksAt = if (changedVote != null) {
-            votesEstimatedUnlocksAt(changedVote)
+    private suspend fun getOnChainReferendaFlow(
+        governanceSource: GovernanceSource,
+        chain: Chain,
+        referendaIds: List<ReferendumId>
+    ): Flow<List<OnChainReferendum>> {
+        return if (referendaIds.size == 1) {
+            governanceSource.referenda.onChainReferendumFlow(chain.id, referendaIds.first())
+                .filterNotNull()
+                .map { listOf(it) }
         } else {
-            votesEstimatedUnlocksAt()
-        }
-
-        return priorUnlocksAt.max(votesEstimatedUnlocksAt)
-    }
-
-    private fun priorUnlocksAt(): BlockNumber {
-        return voting.values.filterIsInstance<Voting.Casting>()
-            .maxOfOrNull { it.prior.unlockAt }.orZero()
-    }
-
-    private fun votesEstimatedUnlocksAt(changedVote: AccountVote.Standard): BlockNumber {
-        val changedVoteMaxLock = claimScheduleCalculator.maxConvictionEndOf(changedVote, onChainReferendum.id)
-
-        val currentVotesExceptChanged = votedReferenda.keys - onChainReferendum.id
-        val currentVotesExceptChangedMaxUnlock = currentVotesExceptChanged.maxOfOrNull {
-            val vote = flattenedVotes.getValue(it)
-
-            claimScheduleCalculator.maxConvictionEndOf(vote, it)
-        }.orZero()
-
-        return changedVoteMaxLock.max(currentVotesExceptChangedMaxUnlock)
-    }
-
-    private fun votesEstimatedUnlocksAt(): BlockNumber {
-        return flattenedVotes.maxOfOrNull { (referendumId, vote) ->
-            claimScheduleCalculator.maxConvictionEndOf(vote, referendumId)
-        }.orZero()
-    }
-
-    private fun Map<TrackId, Voting>.findVotingFor(onChainReferendum: OnChainReferendum): Voting? {
-        val asOngoing = onChainReferendum.status.asOngoingOrNull()
-
-        return if (asOngoing != null) {
-            // fast-path - we have direct access to trackId
-            get(asOngoing.track)
-        } else {
-            // slow path - referendum is completed so have to find by referendumId
-            values.firstOrNull {
-                it is Voting.Casting && onChainReferendum.id in it.votes.keys
-            }
+            flowOf { governanceSource.referenda.getOnChainReferenda(chain.id, referendaIds) }
+                .map { it.values.toList() }
```

Comment:
A potential problem here:
For 1 referendum we observe it from blockchain but for referenda we are able to take it only one time